web/src/filter.ts

```diff
@@ -388,6 +388,35 @@ export class Filter {
         };
     }
 
+    static canonicalize_narrow(narrow: NarrowTerm[]): NarrowTerm[] {
+        // In presence of `with` term without channel or topic terms in the narrow, the
+        // narrow is populated with the channel and toipic terms through this operation,
+        // so that `with` can be used as a standalone operator to target conversation.
+        const with_term = narrow.find((term: NarrowTerm) => term.operator === "with");
+
+        if (with_term) {
+            let channel_term = narrow.find(
+                (term: NarrowTerm) => Filter.canonicalize_operator(term.operator) === "channel",
+            );
+            let topic_term = narrow.find(
+                (term: NarrowTerm) => Filter.canonicalize_operator(term.operator) === "topic",
+            );
+
+            if (!topic_term) {
+                topic_term = {operator: "topic", operand: ""};
+                const with_index = narrow.indexOf(with_term);
+                narrow.splice(with_index, 0, topic_term);
+            }
+
+            if (!channel_term) {
+                channel_term = {operator: "channel", operand: ""};
+                const topic_index = narrow.indexOf(topic_term);
+                narrow.splice(topic_index, 0, channel_term);
+            }
+        }
+        return narrow;
```

Comment:
I added this section in response to the [comment](https://github.com/zulip/zulip/pull/30114#discussion_r1659390541)

Reply:
This approach has a subtle bug: It mutates the original `self._terms`. I'm not sure what impact that has, but it seems dangerous. Making a new object would avoid that problem. The whole architecture feels strange, but to some extent it's working around limitations of the core filter data structure.

Reply:
Added a copy, with better variable names, to fix this:

``` diff
diff --git a/web/src/filter.ts b/web/src/filter.ts
index c0fe85ac9d..177bf61d2e 100644
--- a/web/src/filter.ts
+++ b/web/src/filter.ts
@@ -389,33 +389,36 @@ export class Filter {
         };
     }
 
-    static canonicalize_narrow(narrow: NarrowTerm[]): NarrowTerm[] {
+    static ensure_channel_topic_terms(orig_terms: NarrowTerm[]): NarrowTerm[] {
         // In presence of `with` term without channel or topic terms in the narrow, the
         // narrow is populated with the channel and toipic terms through this operation,
         // so that `with` can be used as a standalone operator to target conversation.
-        const with_term = narrow.find((term: NarrowTerm) => term.operator === "with");
+        const with_term = orig_terms.find((term: NarrowTerm) => term.operator === "with");
 
-        if (with_term) {
-            let channel_term = narrow.find(
-                (term: NarrowTerm) => Filter.canonicalize_operator(term.operator) === "channel",
-            );
-            let topic_term = narrow.find(
-                (term: NarrowTerm) => Filter.canonicalize_operator(term.operator) === "topic",
-            );
+        if (!with_term) {
+            return orig_terms;
+        }
 
-            if (!topic_term) {
-                topic_term = {operator: "topic", operand: ""};
-                const with_index = narrow.indexOf(with_term);
-                narrow.splice(with_index, 0, topic_term);
-            }
+        const updated_terms = [...orig_terms]
+        let channel_term = updated_terms.find(
+            (term: NarrowTerm) => Filter.canonicalize_operator(term.operator) === "channel",
+        );
+        let topic_term = updated_terms.find(
+            (term: NarrowTerm) => Filter.canonicalize_operator(term.operator) === "topic",
+        );
 
-            if (!channel_term) {
-                channel_term = {operator: "channel", operand: ""};
-                const topic_index = narrow.indexOf(topic_term);
-                narrow.splice(topic_index, 0, channel_term);
-            }
+        if (!topic_term) {
+            topic_term = {operator: "topic", operand: ""};
+            const with_index = updated_terms.indexOf(with_term);
+            updated_terms.splice(with_index, 0, topic_term);
+        }
+
+        if (!channel_term) {
+            channel_term = {operator: "channel", operand: ""};
+            const topic_index = updated_terms.indexOf(topic_term);
+            updated_terms.splice(topic_index, 0, channel_term);
         }
-        return narrow;
+        return updated_terms;
     }
 
     /* We use a variant of URI encoding which looks reasonably
```